principles.rst

```diff
@@ -303,3 +309,135 @@ down GHC even when those features are not active. Yet this principle is importan
 
 From `#378`_, slightly generalized.
+
+
+
+GHC stability principles
+--------------------------
+
+The Haskell ecosystem has a built-in tension between stability and innovation.
+
+One the one hand, breaking changes impose heavy costs:
+
+* Users keep having to update their code
+* Library authors keep having to update their code
+* Updating to a later version of a library can in turn force taking on new dependencies.
+
+These difficulties add friction, discourage adoption, and make Haskell an unpredictable investment.
+
+On the other hand, GHC has always sought to be a laboratory for innovation.  If the stability guarantees are too onerous, we risk imposing a different set of costs:
+
+* Volunteers may get discouraged by the hoops they have to jump through to get a change agreed.
+* The language may become stuck in a local optimum, because moving to a better design would require breaking changes.
+* We want Haskell to be a beautiful, elegant language, not one riddled with inconsistencies, grandfathered in simply because fixing the inconsistency would risk some breakage.
+
+We can't navigate this tension with simple all-or-nothing rules.  We have to take each case on its merits, aware of both sets of costs.  Rather than starting the debate from scratch each time, we can have some general rules, or principles, to guide us.  The `GHC base library proposal <https://github.com/haskellfoundation/tech-proposals/blob/main/proposals/accepted/051-ghc-base-libraries.rst>`_ made a start in this direction; this section discusses GHC itself.
+
+
+Assumptions
+~~~~~~~~~~~~
+
+We assume:
+
+* That we have adopted the  `GHC base library proposal <https://github.com/haskellfoundation/tech-proposals/blob/main/proposals/accepted/051-ghc-base-libraries.rst>`_, which establishes the ``ghc-experimental`` package.
+* That we have identified some extensions as "Experimental".  These extensions may well change, and are subject to
+  much weaker stability goals.  `GHC Proposal #601 <https://github.com/ghc-proposals/ghc-proposals/pull/601>`_, describes a plan along these lines, but all we need for this document is a list of Experimental extensions.
+
+Terminology
+~~~~~~~~~~~~~~~
+
+We define a **stable Haskell package** as follows. A stable Haskell package
+
+* Does not use Experimental extensions.
+* Does not use experimental features.  (Examples: use of the Javascript or Wasm back end, builds on non-tier-1 platforms.)
+* Does not rely on explicitly-undefined behaviour. (Example: ``INCOHERENT`` instance selection.)
+* Does not use ``-Werror`` in its default build configuration.
+* Explicitly specifies a language edition (``Haskell98``, ``GHC2021``), in the source code or the build configuration.
+
+Stability (GR1)
+~~~~~~~~~~~~~~~~
+
+**General rule (GR1)**.  *A stable Haskell package P that works with GHC X should continue to work in subsequent releases of GHC, say GHC Y, provided P's direct dependencies work with GHC Y with only minor version bumps.*
```

Comment:
Based on the previous discussion about version bumps and PVP, I think it would better to drop the "with only minor version bumps" part of (GR1), because it isn't well-specified. Moreover we might imagine a package changing API depending on which GHC version is used. So perhaps something like this?

```suggestion
**General rule (GR1)**.  *A stable Haskell package P that works with GHC X should continue to work in subsequent releases of GHC, say GHC Y, provided P's direct dependencies (i) work with GHC Y and (ii) offer compatible APIs on GHC X and GHC Y.*
```

Reply:
>  I think it would better to drop the "with only minor version bumps" part of (GR1), because it isn't well-specified. 

I missed that somehow.  What exactly is not well specified?

Reply:
I'm referring to this discussion (but I had to start a new thread to make a suggested edit): https://github.com/ghc-proposals/ghc-proposals/pull/625#discussion_r1413217877

I think I roughly understand the idea that "with only minor version bumps" is intended to capture, but pinning it down precisely seems surprisingly tricky.